salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,835 @@
+        if not (group and version):
+            raise ValueError(
+                "Malformed 'apiVersion': {} "
+                "(expected format '<group>/<version>')".format(api_version)
+            )
+
+        self._api_version = api_version
+        self._scope = scope
+        self._client = CustomApiClient(
+            group=group,
+            version=version,
+            kind=kind,
+            plural=plural,
+            scope=scope
+        )
+        self._kind = kind
+
+    api_version = property(operator.attrgetter('_api_version'))
+    kind = property(operator.attrgetter('_kind'))
+    scope = property(operator.attrgetter('_scope'))
+    client = property(operator.attrgetter('_client'))
+
+    @property
+    def key(self):
+        # Used for indexing in `KNOWN_CUSTOM_KINDS`
+        return (self.api_version, self.kind)
+
+
+_CUSTOM_KINDS = [
+    CRKindInfo('monitoring.coreos.com/v1', 'Alertmanager',
+               scope='namespaced', plural='alertmanagers'),
+    CRKindInfo('monitoring.coreos.com/v1', 'Prometheus',
+               scope='namespaced', plural='prometheuses'),
+    CRKindInfo('monitoring.coreos.com/v1', 'PrometheusRule',
+               scope='namespaced', plural='prometheusrules'),
+    CRKindInfo('monitoring.coreos.com/v1', 'ServiceMonitor',
+               scope='namespaced', plural='servicemonitors'),
+]
+
+KNOWN_CUSTOM_KINDS = {kind.key: kind for kind in _CUSTOM_KINDS}
+
+
+class _DictWrapper(object):
+    """Wrapper for dynamic attribute access support over a simple dict.
+
+    Implemented by recursively casting all dicts within the original dict as
+    other `_DictWrapper`s, and overloading the `__getattribute__` to fallback
+    on the internal `_fields` for retrieving an attribute.
+
+    Used by `CustomObject` to emulate the other models for `kubernetes.client`.
+
+    >>> example = _DictWrapper({
+    ...     'a': {'b': 'value'},
+    ...     'c': [{'subkey': 'subvalue'}]
+    ... })
+    >>> example.a.b
+    'value'
+    >>> example.c[0].subkey
+    'subvalue'
+    >>> example.unknown
+    Traceback (most recent call last):
+        ...
+    AttributeError: Custom object has no attribute 'unknown'
+    """
+    def __init__(self, fields):
+        self._fields = fields
+
+    @classmethod
+    def from_value(cls, value):
+        if isinstance(value, dict):
+            return cls(value)
+        if isinstance(value, list):
+            return [cls.from_value(val) for val in value]
+        return value
+
+    def __repr__(self):
+        return repr(self._fields)
+
+    def __getattribute__(self, name):
+        try:
+            return super(_DictWrapper, self).__getattribute__(name)
+        except AttributeError:
+            if name in self._fields:
+                return self.from_value(self._fields[name])
+            for key in self._fields:
+                if _convert_attribute_name(key) == name:
+                    return self.from_value(self._fields[key])
```

Comment:
Why do this and not cast the keys before using a `_DictWrapper`? It's highly inefficient IMO, and it confuses the reader as to what the purpose of this `_DictWrapper` is (merely building a custom `__getattribute__` on top of a dict, recursively).

Reply:
The behavior is basically the same as casting it in the `__init__` (like you did before) just here with this approach the `fields` initial argument can be edited through the `CustomObject`
```
toto.metadata.labels.lafrite = 'Titi'
```

Reply:
True I need to add a `__setattr__` in `CustomObject` as well good catch